core/trino-main/src/main/java/io/trino/operator/TableScanOperator.java

```diff
@@ -295,7 +295,11 @@ public Page getOutput()
             if (!dynamicFilter.getCurrentPredicate().isAll()) {
                 operatorContext.recordDynamicFilterSplitProcessed(1L);
             }
-            source = pageSourceProvider.createPageSource(operatorContext.getSession(), split, table, columns, dynamicFilter);
+            ConnectorPageSource pageSource = pageSourceProvider.createPageSource(operatorContext.getSession(), split, table, columns, dynamicFilter);
+            synchronized (this) {
+                // This is synchronised so we can cancel it, see TableScanOperator.cancel
```

Comment:
`volatile` too (or add `@GuardedBy` here too and synchronize usages of `source`)